### ISSUE TITLE
Secrets don't get re-encrypted and stays persisted as a plain text

### ISSUE THREAD

**Issue by php-coder** [openshift/origin#14864]
When user starts to use secrets encryption, the existing secrets should be encrypted after running `oadm migrate storage --include=secrets --confirm`. But it doesn't happen and secrets stay persisted in the plain text.

##### Version
```
oc v3.6.0-alpha.2+b28a75e-471-dirty
kubernetes v1.6.1+5115d708d7
features: Basic-Auth
```

##### Steps To Reproduce
1. start OpenShift without secrets encryption enabled (for example, `sudo $(which openshift) start --public-master=127.0.0.1 `)
2. create a secret (`oc create secret generic foo --from-literal=password=test123`)
3. stop OpenShift
4. create an encryption config:
```yaml
kind: EncryptionConfig
apiVersion: v1
resources:
  - resources:
    - secrets
    providers:
    - aesgcm:
        keys:
        - name: key1
          secret: c2vjcmv0iglzihnly3vyzq==
    - identity: {}
```
5. generate master config (`sudo $(which openshift) start master --public-master=127.0.0.1 --write-config=$PWD/openshift.local.config/master`)
6. edit master config and enable secrets encryption (`sudoedit openshift.local.config/master/master-config.yaml`)
```yaml
kubernetesMasterConfig:
  apiServerArguments:
    experimental-encryption-provider-config:
    - /path/to/encryption.config
```
7. start OpenShift with secrets encryption enabled (`sudo $(which openshift) start --public-master=127.0.0.1 --master-config=$PWD/openshift.local.config/master/master-config.yaml --node-config=$PWD/openshift.local.config/<node>/node-config.yaml`)
8. check that the secret persisted in the plain text (the `ETCDCTL_API=3 etcdctl --cacert=$PWD/openshift.local.config/master/ca.crt --key=$PWD/openshift.local.config/master/master.etcd-client.key --cert=$PWD/openshift.local.config/master/master.etcd-client.crt --endpoints 'https://127.0.0.1:4001' get /kubernetes.io/secrets/default/foo -w fields | grep Value` command should output something like this: `"Value" : "k8s\x00\n\f\n\x02v1\x12\x06Secret\x12\x91\x01\nr\n\x03foo\x12\x00\x1a\adefault\"&/api/v1/namespaces/default/secrets/foo*$6689f2de-566f-11e7-8bdd-507b9d2b16b92\x008\x00B\f\b㞩\xca\x05\x10\xfd\xf6\xce\xd1\x03z\x00\x12\x13\n\bpassword\x12\atest123\x1a\x06Opaque\x1a\x00\"\x00"`)
8. migrate secrets (`oadm migrate storage --include=secrets --confirm`)
9. check that the secret persisted in the encrypted form

##### Current Result
Step 8: `oadm migrate storage` outputs `summary: total=168 errors=0 ignored=0 unchanged=168 migrated=0`
Step 9: the secret persisted in the plain text (output is similar to the Step 7)

##### Expected Result
Step 8: `oadm migrate storage` should print `summary: total=168 errors=0 ignored=0 unchanged=0 migrated=168`
Step 9: the secret should be encrypted and data should starts from the "k8s:enc:aesgcm:v1:key1:" prefix, for example: `"Value" : "k8s:enc:aesgcm:v1:key1:n\x14YG(\x008i\xee\u007f\xce\xd0...`

##### Additional Information
**IMPORTANT**: I've tested with https://github.com/openshift/origin/pull/14836 and https://github.com/openshift/origin/pull/14838 being applied.

My debug observations:
- Inside of [`UpdateResource()`](https://github.com/openshift/origin/blob/d61e8be758ead94289e37d8d5168490325b8f5f9/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go#L849) the flag `wasCreated` is `false`
- Inside of [`GuaranteedUpdate()`](https://github.com/openshift/origin/blob/d61e8be758ead94289e37d8d5168490325b8f5f9/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L301) we're exiting very early, just before transforming data to storage. At this point `origState.stale` flag is `false` **but it should be `true`**
- `origState` with a wrong `stale` field has been returned from [`getStateFromObject`](https://github.com/openshift/origin/blob/d61e8be758ead94289e37d8d5168490325b8f5f9/vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L457-L481) function

@smarterclayton @liggitt @simo5 I couldn't provide a fix at this moment because I don't have enough time and I'm not familiar with rest/etcd code for doing an update. Next week I'll be on PTO and if this blocked issue will be fixed by someone else, I wouldn't mind.

**Comment by php-coder** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[security] [SCCFSI] Encrypt secrets at REST in etcd](https://trello.com/c/yhuQZ3NP/905-security-sccfsi-encrypt-secrets-at-rest-in-etcd)

**Comment by smarterclayton** [comment]
Yeah, the isStale should be true.  This might be the other change I
realized I needed to make to GuaranteedSTorage for the etcd self link
change.

**Comment by php-coder** [comment]
It turned out that it has been fixed by https://github.com/openshift/origin/pull/15001

@smarterclayton Thanks for fixing this but if you  mentioned me at that issue then I wouldn't try to fix it today... :(

**Comment by smarterclayton** [comment]
Heh, forgot you had an open one

On Jul 10, 2017, at 2:03 PM, Vyacheslav Semushin <notifications@github.com>
wrote:

It turned out that it has been fixed by #15001
<https://github.com/openshift/origin/pull/15001>

@smarterclayton <https://github.com/smarterclayton> Thanks for fixing this
but if you mentioned me at that issue then I wouldn't try to fix it
today... :(

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/issues/14864#issuecomment-314186400>,
or mute the thread
